File: tutorials/012_diagnose_a_pipeline/diagnose_a_pipeline.cpp
// Two diagnostic commands: Graph::validate and Run::start_measurement.
//
// Usage:
//   tutorial_012_diagnose_a_pipeline

#include "neat.h"

#include <opencv2/core.hpp>

#include <iostream>
#include <stdexcept>

int main() {
  try {
    cv::Mat rgb(96, 128, CV_8UC3, cv::Scalar(22, 44, 66));
    if (!rgb.isContinuous())
      rgb = rgb.clone();

    simaai::neat::Graph graph;
    simaai::neat::InputOptions in;
    in.format = "RGB";
    in.width = rgb.cols;
    in.height = rgb.rows;
    in.depth = rgb.channels();
    graph.add(simaai::neat::nodes::Input(in));
    graph.add(simaai::neat::nodes::Output());

    // CORE LOGIC
    // STEP validate-graph
    // validate() checks the Graph before build() and prints any caps problems.
    auto report = graph.validate();
    std::cout << "validate.error_code=" << report.error_code << "\n";
    // END STEP

    // STEP run-with-measurement
    // Build a reusable runner and measure the caller-owned workload.
    simaai::neat::RunOptions run_opt;
    run_opt.output_memory = simaai::neat::OutputMemory::Owned;
    auto run = graph.build(std::vector<cv::Mat>{rgb}, run_opt);
    simaai::neat::MeasureOptions measure_opt;
    measure_opt.title = "tutorial 011 diagnosis";
    auto scope = run.start_measurement(measure_opt);
    simaai::neat::TensorList out = run.run(std::vector<cv::Mat>{rgb}, /*timeout_ms=*/1000);
    if (out.empty())
      throw std::runtime_error("missing output tensor");
    const simaai::neat::MeasureReport measured = scope.stop();
    // END STEP

    // STEP read-diagnostics
    // Post-run diagnostics come from the measurement report.
    std::cout << "measure.inputs_enqueued=" << measured.counters.inputs_enqueued
              << " outputs_pulled=" << measured.counters.outputs_pulled << "\n";
    std::cout << "measure.text_size=" << measured.to_text().size() << "\n";
    // END STEP
    // END CORE LOGIC

    std::cout << "[OK] 012_diagnose_a_pipeline\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/012_diagnose_a_pipeline/diagnose_a_pipeline.py
#!/usr/bin/env python3
"""Validate a Graph, measure a Run workload, then read the report.

Usage:
  python3 diagnose_a_pipeline.py
"""
from __future__ import annotations

import argparse
import sys

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )

import numpy as np


def main(argv: list[str]) -> int:
  argparse.ArgumentParser(description=__doc__).parse_args(argv[1:])

  rgb = np.full((96, 128, 3), 77, dtype=np.uint8)
  tensor = pyneat.Tensor.from_numpy(rgb, copy=True, image_format=pyneat.PixelFormat.RGB)

  inp = pyneat.InputOptions()
  inp.format = pyneat.Format.RGB
  inp.width = 128
  inp.height = 96
  inp.depth = 3

  graph = pyneat.Graph()
  graph.add(pyneat.nodes.input(inp))
  graph.add(pyneat.nodes.output())

  # CORE LOGIC
  # STEP validate-graph
  # Validate pipeline before building.
  report = graph.validate()
  print(f"validate_error_code={report.error_code}")
  # END STEP

  # STEP run-with-measurement
  # Build a reusable runner and measure the caller-owned workload.
  ropt = pyneat.RunOptions()
  ropt.output_memory = pyneat.OutputMemory.Owned
  run = graph.build([tensor], ropt)
  measure = pyneat.MeasureOptions()
  measure.title = "tutorial 011 diagnosis"
  scope = run.start_measurement(measure)
  run.run([tensor], timeout_ms=1000)
  measured = scope.stop()
  # END STEP

  # STEP read-diagnostics
  # Read diagnostics from the measurement report.
  print(f"inputs_enqueued={measured.counters.inputs_enqueued} outputs_pulled={measured.counters.outputs_pulled}")
  print(f"measure_text_size={len(measured.to_text())}")
  # END STEP
  # END CORE LOGIC
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
